runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java

```diff
@@ -133,6 +134,43 @@ private void validateStateOfWrittenEntry(long address, @Nonnull ILogData ld) {
         }
     }
 
+    public void sparseTrim(@NonNull IToken token, @NonNull ILogData trimInfo) {
+        // TODO(Xin): should sparseTrim should be epoch-aware?
+        log.info("sparse trim[{}]", token.getSequence());
+        final int numRetries = 3;
+
+        for (int x = 0; x < numRetries; x++) {
+            try {
+                layoutHelper(e -> {
+                    trimInfo.useToken(token);
+                    trimInfo.setId(runtime.getParameters().getClientId());
+                    e.getLayout()
+                            .getSegment(token.getSequence())
+                            .getStripes().stream()
+                            .flatMap(strip -> strip.getLogServers().parallelStream())
```

Comment:
Don't use parallelStream, it uses the common pool, which affects other workloads.
Also you probably could use `e.getLogUnitClient(address, index)`